### DISCUSSION TITLE
✨ Pre-Releases Supporting Uniswap V4 !! ✨

### DISCUSSION THREAD

**Discussion by Elnaril** [Elnaril/uniswap-universal-router-decoder#82] [Announcements]
This library starts to support Uniswap V4 with this alpha version that has just been released. Here is the release note:

### v2.0.0a1
 - Add support for some V4 functions and features:
   - `V4_INITIALIZE_POOL`
   - `V4_POSITION_MANAGER_CALL`
     - `MINT_POSITION`
     - `SETTLE`
     - `SETTLE_PAIR`
     - `CLOSE_CURRENCY`
     - `UNWRAP`
   - `V4_SWAP`
     - `SWAP_EXACT_IN_SINGLE`
     - `SETTLE`
     - `SETTLE_ALL`
     - `TAKE_ALL`
   - Pool Key and Pool Id encoding
 - Add support for `PERMIT2_TRANSFER_FROM`
 - Custom contract error decoding
 - Encoding refactoring

More V4 functions will be added in following alpha/beta versions before the release of the stable version, which should happen shortly after Uniswap V4 and the Universal Router are deployed on Mainnet and Unichain.  
Expect also some heavy refactoring and breaking changes with these alpha release.

---

Feel free to give me any feedback on this release here, or open a discussion or a ticket about a feature that should be in the next one!

**Comment by Elnaril** [comment]
### v2.0.0a2
More Uniswap V4 function supported in this 2nd pre-release!

- Add support for more V4 functions and features:
  - WRAP & fix UNWRAP
  - SWAP_EXACT_IN
  - SWAP_EXACT_OUT_SINGLE
- Remove support for Python 3.8
- More refactoring
- Update repository picture

**Reply by galinaggil80** [reply]
thx! rly!

**Reply by galinaggil80** [reply]
tranks!

**Comment by Vlada3333** [comment]
Excellent, thanks!

**Comment by Elnaril** [comment]
### v2.0.0a3
Add the remaining V4 Functions supported by the Universal Router:

- CLEAR_OR_TAKE
- TAKE
- TAKE_PAIR
- TAKE_PORTION

**Comment by Nika250** [comment]
thanks for the information!

**Comment by Elnaril** [comment]
### v2.0.0b1 - Fix issues and add doc

- Fix SWAP_EXACT_IN and SWAP_EXACT_OUT issue
- Add/improve documentation
- Update default router address to new Mainnet's one
- Minor refactoring

**Comment by johann75sch** [comment]
Is there an example somewhere of how to use ERC-6909?
